crates/bevy_reflect/src/func/registry.rs

```diff
@@ -0,0 +1,233 @@
+use alloc::borrow::Cow;
+use core::fmt::Debug;
+use std::sync::{Arc, PoisonError, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use bevy_utils::HashMap;
+
+use crate::func::{DynamicFunction, FunctionRegistrationError, IntoFunction};
+
+/// A registry of [reflected functions].
+///
+/// This is the function-equivalent to the [`TypeRegistry`].
+///
+/// [reflected functions]: crate::func
+/// [`TypeRegistry`]: crate::TypeRegistry
+#[derive(Default)]
+pub struct FunctionRegistry {
+    /// Maps function [names] to their respective [`DynamicFunctions`].
+    ///
+    /// [names]: DynamicFunction::name
+    /// [`DynamicFunctions`]: DynamicFunction
+    functions: HashMap<Cow<'static, str>, DynamicFunction>,
+}
+
+impl FunctionRegistry {
+    /// Attempts to register the given function if it has not yet been registered already.
+    ///
+    /// Functions are mapped according to their [name].
+    /// If a `DynamicFunction` with the same name already exists, it will not be registered again,
+    /// and an error will be returned.
+    /// To register the function anyway, overwriting any existing registration, use [`overwrite_registration`] instead.
+    ///
+    /// [name]: DynamicFunction::name
+    /// [`overwrite_registration`]: Self::overwrite_registration
+    pub fn register<F, Marker>(
+        &mut self,
+        function: F,
+    ) -> Result<&mut Self, FunctionRegistrationError>
+    where
+        F: IntoFunction<Marker> + 'static,
+    {
+        let function = function.into_function();
+        let name = function.name().clone();
+        self.functions
+            .try_insert(name, function)
+            .map_err(|err| FunctionRegistrationError::DuplicateName(err.entry.key().clone()))?;
+
+        Ok(self)
+    }
+
+    /// Registers the given function, overwriting any existing registration.
+    ///
+    /// Functions are mapped according to their [name].
+    /// To avoid overwriting existing registrations, it's recommended to use the [`register`] method instead.
+    ///
+    /// [name]: DynamicFunction::name
+    /// [`register`]: Self::register
+    pub fn overwrite_registration<F, Marker>(&mut self, function: F)
+    where
+        F: IntoFunction<Marker> + 'static,
+    {
+        let function = function.into_function();
+        let name = function.name().clone();
+        self.functions.insert(name, function);
+    }
+
+    /// Get a reference to a registered function by [name].
+    ///
+    /// [name]: DynamicFunction::name
+    pub fn get(&self, name: &str) -> Option<&DynamicFunction> {
+        self.functions.get(name)
```

Comment:
TIL that this works (the key type is different from the supplied one). Nifty!

Reply:
Yeah `get` and friends just require `K: Borrow<Q>` where `K` is the actual key type and `Q` is the supplied one (and `Cow<'static, str>: &str`)